scRNA-seq-advanced/03-dataset_integration.Rmd

```diff
@@ -149,49 +181,109 @@ sce_list
+rownames( colData(sce_list$SCPCL000478) ) %>%
+  # Only print out the first 6 for convenience
+  head()
+```
+
+Because barcodes theoretically can be repeated across libraries, we could end up in a situation with non-unique rownames, which would make R complain!
+
+One way to ensure that barcodes remain unique even after merging is to actually modify the rownames by _prepending_ the relevant library name. 
+For example, we'll update the `CCTTTGGCACTACCCT` in the `SCPCL000478` SCE to become: `SCPCL000478-CCTTTGGCACTACCCT`, thereby ensuring fully unique rownames for all cells.
+
+#### Preserving library information at the gene level
+
+The `rowData` table in SCE objects will often contain both "general" and "library-specific" information, for example:
+
+
+```{r rowdata colnames}
+head( rowData(sce_list$SCPCL000478) )
+```
+
+Here, the rownames are Ensembl gene IDs, and columns are `gene_symbol`, `mean`, and `detected`. 
+The `gene_symbol` column is general information not specific to any library or experiment, but `mean` and `detected` are library-specific gene statistics. 
+So, `gene_symbol` does not need to be traced back to its originating library, but `mean` and `detected` do.
+To this end, we can take a similar approach to what we'll do for cell barcodes: We can change the library-specific `rowData` column names by prepending the library name.
+For example, rather than being called `mean`, this column will be named `SCPCL000478-mean` for the `SCPCL000478` library.
+
+All our SCE objects have the same `rowData` columns (as we can see in the next chunk), so we'll perform this renaming across all SCEs.
+
+```{r compare rowdata}
+# Use `purrr::map()` to quickly extract rowData names for all SCEs
+purrr::map(sce_list,
+           # This syntax is a formula where the period is a stand-in for 
+           # each item in sce_list
+           ~names(rowData(.)))
+```
 
 
-First, let's determine what the shared genes among SCE objects are using some `purrr` magic.
-Ultimately, we'll filter SCEs down to contain only these genes, thereby ensuring that their genes match.
+#### Ensuring that only shared genes are used
+
+The final step in ensuring SCE compatibility is to make sure they all contain the same genes, which are unambiguously stored in our SCE objects as Ensembl ID `rowData` rownames, which you can map back to gene names by simply asking my co-instructor Josh. 
+
+We can use some `purrr` magic to quickly find the set of shared genes among our libraries:
 ```{r shared genes}
 shared_genes <- sce_list %>%
   # get rownames for each SCE in sce_list
   purrr::map(rownames) %>%
   # reduce to the _intersection_ among lists
   purrr::reduce(intersect)
 
-# Vector of genes that are present across all SCEs
+# We've now created a vector genes that are present in all SCEs
 head(shared_genes)
 ```
 
+<!--- I might get rid of this text and chunk `check shared genes` for time/space? --> 
```

Comment:
I think you can get rid of this.

Reply:
I realized we can keep a much simpler version!

```
# The number of genes in an SCE corresponds to its number of rows:
sce_list %>%
  purrr::map(nrow)
```